{% extends "base.html" %}

{% block maincontent %}
<div class="error">
  {% for mesg in get_flashed_messages() %}
  <p style="color:pink;">{{ mesg }}</p>
  {% endfor %}
</div>

<div class="submission">

  <form action="" method="post" novalidate enctype="multipart/form-data">
    {{ form.hidden_tag() }}
    <p>
      Upload .sqlite file:
      {{ form.user_upload}}
    </p>
    <p>
      Upload .csv with query:
      {{ form.user_upload_query }}
    </p>
    <p>
      Minimum occurences to consider for genotypes:
      {{ form.user_min_occurences}}
      <br /><i>(Recommended: 3)</i>
    </p>

    <p>
      Quantize gaps to:
      {{ form.user_min_gap_size }} base pairs
      <br /><i>(Recommended: 500)</i>
    </p>
    <p style="text-align:center;">
      {{ form.submit() }}
    </p>
  </form>

</div>

<div class="infoBlurb">
  <h2>What it does:</h2>
  <p>
    This program analyzes the output from <a href="https://efi.igb.illinois.edu/efi-gnt">EFI-GNT</a> and organizes it in a way which might be useful to researchers studying microbial natural products.
  </p>

  <p>
    Gene clusters are grouped together based on the order and position of the enzymes they encode. The groupings can then be mapped onto the cytoscape network normally used to visualize <a href="https://efi.igb.illinois.edu/efi-gnt">EFI-GNT</a> jobs - except now, instead of colouring nodes by organism genus, they can be coloured based on the genotypes classified by this program.
  </p>

  <h4><a href="static/example_report.html">Here's </a>a sample report.</h4>

  <h2>How it works:</h2>
  <p>
    First, the user inputs a list of PFAMs and IPRO codes which they would like to analyze. For example:
  </p>
  <pre><code>
    Transporters: PF11111, IPRO212
    Methylases:   PF22222, PF33333, PF44444
    Oxidases:     PF66666, PF77777
  </code></pre>
  <p>
    Next, the user uploads the <code>.sqlite</code> file from an EFI-GNT job (the one used by EFI-GNT to 'View Saved Diagrams'). This file contains all the raw data from the EFI-GNT job.
  </p>
  <p>
    Cytotyper will parse all the gene clusters in the <code>.sqlite</code> file and label genes which match any of the Pfams/IPROs specified by the user:
  </p>
  <img src='static/diagrams/fig1.png' />
  <p>
    Next, Cytotyper fills out the "un-queried" portions of the gene sequences with gap units. The length of these gap units is specified by the user, though I find a number between 300-500 base pairs works well. Gaps are represented by the number of base pairs between two queried enzymes, divided by the length of a gap unit - rounded down:
  </p>
  <img src='static/diagrams/fig2.png' />
  <p>
    The 'distilled' representations of the gene clusters are then grouped and labelled:
  </p>
  <img src='static/diagrams/fig3.png' />
  <p>
    Finally, the group labellings are exported as a <code>.csv</code> file which can be super-imposed onto the genomic neighhborhood network generated by EFI-GNT:
  <img src='static/diagrams/fig4.png' />
  </p>
  <h2>How to use it:</h2>
  <p>
    <b>1.</b> Obtain the <code>.sqlite</code> file for an EFI-GNT job.
  </p>
  <p>
    <b>2.</b> Using your favourite spreadsheet program, prepare a <code>query.csv</code> that looks like this:
  </p>
  <img src='static/diagrams/query.png' />
  Note! It's very important that the first row ("Labels" and "Pfams") are exactly as above! <a href="static/example_query.csv">Here's a template you can use. </a>
  <p>
    <b>3.</b> Input the parameters and hit GO!
  </p>
</div>

{% endblock %}
